### PR TITLE
Restructure the project

### LINKED ISSUE #590: Source restructuring
Create new source layout for easier packaging and restructure modules to subdirectories.

### PR DESCRIPTION
**Pull request by whisperity**
In this PR, we introduce a new internal structure for CodeChecker in anticipation of further modularisation.

 * The many different libraries, such as `viewer_clients` and `report_server` are condensed in a **single** `lib/` structure.
    * The `libcodechecker` structure is as follows:
       * Every "module" of _CodeChecker_ (logger, analyzer, storage, cmdline-client) is now given its own subfolder/sublibrary.
       * Shared libraries are in the root of `libcodechecker`
       * (Some cross-references between libraries remain which should be ironed out in the future, marked with a `TODO` in the code.)
    * This rename makes it easier for IDEs such as PyCharm to properly resolve references between parts of the program.
 * The build and install script has been modified accordingly.
 * The behaviour of _CodeChecker_ is **not** modified at all: the same entry-points and invocation rules apply.
 * Generated Thrift Python codes are stored in `gencodechecker` which is included as a top-level directory in the `PYTHONPATH`.
 * The developer folder structure now follows conventions such as `bin`, `libcodechecker`, `vendor` instead of names unique to our project.

Resolves #590

### CODE REVIEW

File: bin/CodeChecker.py [hunk 1]
```diff
@@ -17,12 +17,12 @@
...
 import shared
 
-from cmdline_client import cmd_line_client
-from codechecker_lib import arg_handler
-from codechecker_lib import util
-from codechecker_lib import logger
-from codechecker_lib.logger import LoggerFactory
-from codechecker_lib.analyzers import analyzer_types
+from lib.cmd import cmd_line_client
+from lib import arg_handler
```
Comment (gyorb) [inline bin/CodeChecker.py]: I would prefer some more descriptive name like `libcodechecker`, I know this is not that short but still.

File: lib/client.py [hunk 2]
```diff
@@ -12,17 +12,19 @@
...
-from DBThriftAPI.ttypes import SuppressBugData
 from thrift import Thrift
 from thrift.protocol import TBinaryProtocol
 from thrift.transport import TSocket
 from thrift.transport import TTransport
 
-from codechecker_lib import suppress_file_handler
-from codechecker_lib.logger import LoggerFactory
-from storage_server import report_server
+from api import shared
```
Comment (gyorb) [inline lib/client.py]: The same here for the `api`, can you think of a more descriptive name? It is a bit too general.

**Comment by gyorb** [comment]
Please remove the [ci skip] from the commit messages. I know the tests will fail because of the restructuring and renaming but the package building should be successful.
If #583 will be accepted, the tests can be fixed to work with the new layout.

File: libcodechecker/analyze/analysis_manager.py [hunk 3]
```diff
@@ -16,9 +16,9 @@
 import traceback
 from collections import defaultdict
 
-from codechecker_lib import analyzer_env
-from codechecker_lib.logger import LoggerFactory
-from codechecker_lib.analyzers import analyzer_types
+from libcodechecker.analyze import analyzer_env
```
Comment (gyorb) [inline libcodechecker/analyze/analysis_manager.py]: Could we group/sort the imports from the same modules?

File: libcodechecker/analyze/analyzer.py [hunk 4]
```diff
@@ -14,12 +14,12 @@
...
 
-from codechecker_lib import analysis_manager
-from codechecker_lib import client
-from codechecker_lib import skiplist_handler
-from codechecker_lib import analyzer_env
-from codechecker_lib.logger import LoggerFactory
-from codechecker_lib.analyzers import analyzer_types
+from libcodechecker.analyze import analysis_manager
+from libcodechecker import client
+from libcodechecker.analyze import skiplist_handler
```
Comment (gyorb) [inline libcodechecker/analyze/analyzer.py]: The same import sorting here.

**Comment by whisperity** [comment]
@gyorb I did the import sorting on all `libcodechecker` files, please check and yell if I left something out.
I tried to follow the convention below, which I think is a good convention for us to use from this point now on.

From top to bottom, imports are ordered, as follows:

 1. System-wide imports
 2. External module imports (`from thrift.Thrift import yaddayadda`)
 3. API imports (`import shared` and `from codecheckerDBapi import ttypes`)
 4. `from libcodechecker import arg_handler` (top-level shared lib import)
    a. Imports from a module in the shared folder shares alphabetical ordering on this level. So it's
        i. `from libcodechecker import arg_handler`
        ii. `from libcodechecker.logger import LoggerFactory` (alphabetically sorted as something with `l`)
        iii. `from libcodechecker import zzzz_something`
 5. Imports from module-specifics (`from libcodechecker.analyzer import analyzer_types`)
    a. These imports are sorted alphabetically on the module name: `analyzer`, `cmd`, `log`, `storage`
    b. In case there is a smaller entity imported `from libcodechecker.analyzer.SomethingAnalyzer import SomeStuff`), the 4a. rule applies.

File: config/package_layout.json [hunk 5]
```diff
@@ -18,13 +18,9 @@
...
     "ld_logger": "ld_logger",
     "codechecker_modules": "cc_lib/python2.7",
-    "codechecker_lib": "cc_lib/python2.7/codechecker_lib",
-    "codechecker_gen": "cc_lib/python2.7/codechecker_gen",
-    "codechecker_db_model": "cc_lib/python2.7/db_model",
-    "codechecker_db_migrate": "cc_lib/python2.7/db_migrate",
-    "storage_server_modules": "cc_lib/python2.7/storage_server",
-    "viewer_server_modules": "cc_lib/python2.7/viewer_server",
-    "cmdline_client": "cc_lib/python2.7/cmdline_client"
+    "codechecker_lib": "lib/python2.7/libcodechecker",
```
Comment (gyorb) [inline config/package_layout.json]: I think we can rename the `codechecker_lib` to `libcodechecker` and `codechecker_gen` to `gencodechecker` here in the json. Please update the package builder too!

**Comment by gyorb** [comment]
We should write the import order in a `CONTRIBUTING.md` guide under a python style guide section.

File: libcodechecker/cmd/cmd_line_client.py [hunk 6]
```diff
@@ -163,6 +163,8 @@ def setupClient(host, port, uri):
                     print("Authenticated using pre-configured credentials...")
                 except shared.ttypes.RequestFailed:
                     print_err = True
+            else:
```
Comment (gyorb) [inline libcodechecker/cmd/cmd_line_client.py]: Please move the changes non relevant to the restructuring into a separate commit

**Pull request merged**